spring-tx/src/main/java/org/springframework/dao/support/DataAccessUtils.java

```diff
@@ -56,6 +60,89 @@
 		return results.iterator().next();
 	}
 
+	/**
+	 * Return a single result object from the given Stream.
+	 * <p>Returns {@code null} if 0 result objects found;
+	 * throws an exception if more than 1 element found.
+	 * @param results the result Stream (can be {@code null})
+	 * @return the single result object, or {@code null} if none
+	 * @throws IncorrectResultSizeDataAccessException if more than one
+	 * element has been found in the given Stream
+	 */
+	@Nullable
+	public static <T> T singleResult(@Nullable Stream<T> results) throws IncorrectResultSizeDataAccessException {
+		if (results == null) {
+			return null;
+		}
+		try (results) {
+			List<T> resultList = results.limit(2).toList();
+			if (resultList.size() > 1) {
+				throw new IncorrectResultSizeDataAccessException(1, resultList.size());
```

Comment:
I think IncorrectResultSizeDataAccessException should contain the size of the results stream.
if `resultList.size()` will be used, actual size of IncorrectResultSizeDataAccessException is `<=2` always unlike actual size of results.

If prefer stream way, How about utilze `teeing`?
```java
		results.collect(teeing(counting(), reducing((first, second) -> first), (count, result) -> {
			if (count > 1) {
				throw new IncorrectResultSizeDataAccessException(1, count.intValue());
			}
			return result.orElse(null);
		}))
```

Reply:
Just a suggestion. There are multiple constructors provided by `IncorrectResultSizeDataAccessException` class. 

* One of the constructors doesn't take `actualSize` argument and assigns the value `-1` if the size if unknown. You can consider using it
  * `public IncorrectResultSizeDataAccessException(int expectedSize)` - [Javadoc](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/dao/IncorrectResultSizeDataAccessException.html#IncorrectResultSizeDataAccessException-int-)
* Another one explicitly suggests setting the value to `-1` if size is unknown
  * `public IncorrectResultSizeDataAccessException(int expectedSize, int actualSize)` - [Javadoc](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/dao/IncorrectResultSizeDataAccessException.html#IncorrectResultSizeDataAccessException-int-int-)

Both of them would be suitable for this scenario since size is unknown

Reply:
It makes sense.

Reply:
```suggestion
				throw new IncorrectResultSizeDataAccessException(1);
```

Adding as suggestion for convenience

Reply:
yeah let's use the `public IncorrectResultSizeDataAccessException(int expectedSize)` for the iterator and stream method. 
What do you mean by "UT"?